### PR TITLE
added ability to restrict addons-nginx-ingress with loadBalancerSourc…

### LINKED ISSUE #269: Ability to restrict addons-nginx-ingress LoadBalancer Source Ranges
In many of our Shoots, our customers are deploying test systems and mostly want to limit access to the exposed applications from certain ip ranges.

This seems to be implemented using the `loadBalancerSourceRanges` field in Service definition: https://kubernetes.io/docs/tasks/access-application-cluster/configure-cloud-provider-firewall/

Example:
```
kind: Service
metadata:
  name: addons-nginx-ingress-controller
  namespace: kube-system
spec:
  type: LoadBalancer
  loadBalancerSourceRanges:
  - 1.1.1.0/24
```

### PR DESCRIPTION
**Pull request by cihangirbesiktas**
…eRanges regarding issue-269

**What this PR does / why we need it**:
This PR adds the ability to pass list of source ranges to the addon chart that deploy addons-nginx-ingress-controller so that it will be possible to restrict access to the exposed services in shoot clusters.

**Which issue(s) this PR fixes**:
Fixes #269

**Special notes for your reviewer**:

**Release note**:

```improvement user
Ability to configure loadBalancerSourceRanges for the service resource in .spec.addons.nginx-ingress
```

### CODE REVIEW

File: pkg/apis/garden/v1beta1/types.go [hunk 1]
```diff
@@ -870,6 +870,8 @@ type ClusterAutoscaler struct {
 // NginxIngress describes configuration values for the nginx-ingress addon.
 type NginxIngress struct {
 	Addon `json:",inline"`
+	// LoadBalancerSourceRanges is list of whitelist IP sources for NginxIngress
```
Comment (mvladev) [inline pkg/apis/garden/v1beta1/types.go:873]: optional fields must have a `// +optional` tag
Check https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#optional-vs-required for more info

**Comment by cihangirbesiktas** [comment]
I tested the changes on a shoot cluster created in Azure, and it successfully configured `loadBalancerSourceRanges` of the `addons-nginx-ingress-controller` service resource.

**Comment by luisdavim** [comment]
this fixes #269

File: pkg/openapi/openapi_generated.go [hunk 2]
```diff
@@ -3796,6 +3810,11 @@ func schema_pkg_apis_garden_v1beta1_Seed(ref common.ReferenceCallback) common.Op
 					},
 				},
 			},
+			VendorExtensible: spec.VendorExtensible{
+				Extensions: spec.Extensions{
+					"x-kubernetes-print-columns": "custom-columns=NAME:.metadata.name,DOMAIN:.spec.ingressDomain,CLOUDPROFILE:.spec.cloud.profile,REGION:.spec.cloud.profile,READY:.status.conditions[?(@.type == 'Available')].status",
+				},
+			},
```
Comment (bergerx) [inline pkg/openapi/openapi_generated.go]: This doesn't seem relevant to the PR.

This question is more to the owners, any idea about why this was missing before?

File: pkg/apis/garden/types.go [hunk 3]
```diff
@@ -891,6 +891,8 @@ type ClusterAutoscaler struct {
 // NginxIngress describes configuration values for the nginx-ingress addon.
 type NginxIngress struct {
 	Addon
+    // LoadBalancerSourceRanges is list of whitelist IP sources for NginxIngress
+	LoadBalancerSourceRanges []string
```
Comment (bergerx) [inline pkg/apis/garden/types.go]: Can you fix the indent of the comment?

File: pkg/openapi/openapi_generated.go (hunk 2 above)
Comment (mvladev) [reply pkg/openapi/openapi_generated.go]: OpenAPI generation was broken in k8s 1.11. I've made a PR to fix it in #295

File: pkg/apis/garden/types.go (hunk 3 above)
Comment (mvladev) [reply pkg/apis/garden/types.go]: You should run `go fmt ./...`

**Review by bergerx — APPROVED** [review]

**Review by mvladev — CHANGES_REQUESTED** [review]
Looks good. Just a small change for `// +optional` tag is required

**Review by mvladev — APPROVED** [review]
/lgtm @dkistner can you have a look?

File: pkg/operation/botanist/addons.go [hunk 4]
```diff
@@ -43,7 +43,22 @@ func (b *Botanist) DestroyIngressDNSRecord() error {
 // GenerateNginxIngressConfig generates the values which are required to render the chart of
 // the nginx-ingress-controller properly.
 func (b *Botanist) GenerateNginxIngressConfig() (map[string]interface{}, error) {
```
Comment (rfranzke) [inline pkg/operation/botanist/addons.go]: Any change here is not needed (actually, can you please remove the whole method (it is not called anywhere and a left over of previous versions)).

File: pkg/operation/cloudbotanist/awsbotanist/addons.go [hunk 5]
```diff
@@ -151,6 +151,9 @@ func (b *AWSBotanist) GenerateNginxIngressConfig() (map[string]interface{}, erro
 			"config": map[string]interface{}{
 				"use-proxy-protocol": "true",
 			},
+			"service": map[string]interface{}{
```
Comment (rfranzke) [inline pkg/operation/cloudbotanist/awsbotanist/addons.go]: As transporting the `loadBalancerSourceRanges` to the Helm chart seems to be not cloud specific, please move this part to https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/addons.go#L147 and keep the cloud specific implementations untouched.

File: pkg/operation/cloudbotanist/azurebotanist/addons.go [hunk 6]
```diff
@@ -68,5 +68,11 @@ func (b *AzureBotanist) GenerateAdmissionControlConfig() (map[string]interface{}
 
 // GenerateNginxIngressConfig generates values which are required to render the chart nginx-ingress properly.
 func (b *AzureBotanist) GenerateNginxIngressConfig() (map[string]interface{}, error) {
-	return common.GenerateAddonConfig(nil, b.Shoot.NginxIngressEnabled()), nil
+	return common.GenerateAddonConfig(map[string]interface{}{
```
Comment (rfranzke) [inline pkg/operation/cloudbotanist/azurebotanist/addons.go]: As transporting the `loadBalancerSourceRanges` to the Helm chart seems to be not cloud specific, please move this part to https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/addons.go#L147 and keep the cloud specific implementations untouched.

File: pkg/operation/cloudbotanist/gcpbotanist/addons.go [hunk 7]
```diff
@@ -57,5 +57,11 @@ func (b *GCPBotanist) GenerateAdmissionControlConfig() (map[string]interface{},
 
 // GenerateNginxIngressConfig generates values which are required to render the chart nginx-ingress properly.
 func (b *GCPBotanist) GenerateNginxIngressConfig() (map[string]interface{}, error) {
-	return common.GenerateAddonConfig(nil, b.Shoot.NginxIngressEnabled()), nil
+	return common.GenerateAddonConfig(map[string]interface{}{
```
Comment (rfranzke) [inline pkg/operation/cloudbotanist/gcpbotanist/addons.go]: As transporting the `loadBalancerSourceRanges` to the Helm chart seems to be not cloud specific, please move this part to https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/addons.go#L147 and keep the cloud specific implementations untouched.

File: pkg/operation/cloudbotanist/localbotanist/addons.go [hunk 8]
```diff
@@ -52,5 +52,11 @@ func (b *LocalBotanist) GenerateAdmissionControlConfig() (map[string]interface{}
 
 // GenerateNginxIngressConfig generates values which are required to render the chart nginx-ingress properly.
 func (b *LocalBotanist) GenerateNginxIngressConfig() (map[string]interface{}, error) {
-	return common.GenerateAddonConfig(nil, b.Shoot.Info.Spec.Addons.NginxIngress.Enabled), nil
```
Comment (rfranzke) [inline pkg/operation/cloudbotanist/localbotanist/addons.go:55]: As transporting the `loadBalancerSourceRanges` to the Helm chart seems to be not cloud specific, please move this part to https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/addons.go#L147 and keep the cloud specific implementations untouched.

File: pkg/operation/cloudbotanist/openstackbotanist/addons.go [hunk 9]
```diff
@@ -50,5 +50,11 @@ func (b *OpenStackBotanist) GenerateAdmissionControlConfig() (map[string]interfa
 
 // GenerateNginxIngressConfig generates values which are required to render the chart nginx-ingress properly.
 func (b *OpenStackBotanist) GenerateNginxIngressConfig() (map[string]interface{}, error) {
-	return common.GenerateAddonConfig(nil, b.Shoot.NginxIngressEnabled()), nil
```
Comment (rfranzke) [inline pkg/operation/cloudbotanist/openstackbotanist/addons.go]: As transporting the `loadBalancerSourceRanges` to the Helm chart seems to be not cloud specific, please move this part to https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/addons.go#L147 and keep the cloud specific implementations untouched.

**Review by rfranzke — CHANGES_REQUESTED** [review]
I made a few suggestions, please take a look and address them. After doing that, please squash your commits to get the PR ready to be merged.

**Comment by rfranzke** [comment]
Thanks. Looks ready **after** rebasing and squashing.

**Review by rfranzke — APPROVED** [review]
LGTM

**Pull request merged**